### PR TITLE
Validate that GBK coders are always set correctly.

### PR DESCRIPTION
**Pull request by robertwb**
------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by robertwb** [comment]
R: @lukecwik

**Review by lukecwik — COMMENTED** [review]
Can we remove / consolidate any of the GBK coder logic that we have, e.g:
```
dataflow_runner.py#_verify_gbk_coders
common.py#group_by_key_input_visitor (BEAM-115)
```

File: sdks/python/apache_beam/runners/common.py [hunk 1]
```diff
@@ -1852,3 +1853,46 @@ def visit_transform(self, transform_node):
...
+      if len(transform_proto.outputs) != 1:
+        raise ValueError("Unexpected number of outputs: %s" % transform_proto)
+      input_coder = get_coder(next(iter(transform_proto.inputs.values())))
+      output_coder = get_coder(next(iter(transform_proto.outputs.values())))
+      if input_coder.spec.urn != common_urns.coders.KV.urn:
+        raise ValueError(
+            "Bad coder for input of %s: %s" % (transform_id, input_coder))
+      if output_coder.spec.urn != common_urns.coders.KV.urn:
+        raise ValueError(
+            "Bad coder for output of %s: %s" % (transform_id, input_coder))
```
Comment (lukecwik) [inline sdks/python/apache_beam/runners/common.py:1882]:
```suggestion
            "Bad coder for output of %s: %s" % (transform_id, output_coder))
```

**Comment by robertwb** [comment]
Thanks for the error message fixes. I also found and fixed another case. 

As for redundancy, `group_by_key_input_visitor` is what coerces them, this is a check that in the end (at the proto level) all is well. `DataflowRunner._verify_gbk_coders` runs as part of v1beta3 translation.

**Review by lukecwik — CHANGES_REQUESTED** [review]
Your change looks good but it picked up a lot of cases where GBK is not well formed it seems.

One example seems to be that we force a deterministic key coder for the input but don't use it for the output,
```
E       ValueError: Incompatible input coder ref_AppliedPTransform_assert_that-Group-CoGroupByKeyImpl-GroupByKey_31 and output coder spec {
E         urn: "beam:coder:kv:v1"
E       }
E       component_coder_ids: "ref_Coder_DeterministicFastPrimitivesCoder_6"
E       component_coder_ids: "ref_Coder_FastPrimitivesCoder_7"
E        for transform spec {
E         urn: "beam:coder:kv:v1"
E       }
E       component_coder_ids: "ref_Coder_FastPrimitivesCoder_7"
E       component_coder_ids: "ref_Coder_IterableCoder_9"
```

File: sdks/python/apache_beam/runners/common.py [hunk 2]
```diff
@@ -1852,3 +1853,46 @@ def visit_transform(self, transform_node):
...
+  def get_coder(pcoll_id):
+    return pipeline_proto.components.coders[
+        pipeline_proto.components.pcollections[pcoll_id].coder_id]
+
+  def validate_transform(transform_id):
+    transform_proto = pipeline_proto.components.transforms[transform_id]
+
+    # Currently the only validation we perform is that GBK operations have
+    # their coders set properly.
+    if transform_proto.spec.urn == common_urns.primitives.GROUP_BY_KEY.urn:
```
[the remark this answers is not in the document]
Comment (robertwb) [reply sdks/python/apache_beam/runners/common.py:1870]: I don't think we have that property visible at the FnAPI level.

[the remark this answers is not in the document]
Comment (lukecwik) [reply sdks/python/apache_beam/runners/common.py:1870]: Would it be worth converting the coder proto to the coder object and check it?

[the remark this answers is not in the document]
Comment (robertwb) [reply sdks/python/apache_beam/runners/common.py:1870]: I don't think that's worth doing here, and it's not always even possible (e.g. cross-langauge). We could consider elevating the notion of determinism to the model level but that's a larger change.

[the remark this answers is not in the document]
Comment (lukecwik) [reply sdks/python/apache_beam/runners/common.py:1870]: Sounds good.

**Comment by robertwb** [comment]
Run Python PreCommit

**Comment by robertwb** [comment]
Run Python PreCommit

**Comment by robertwb** [comment]
Looks like all the tests are passing. PTAL.

**Review by lukecwik — APPROVED** [review]

**Pull request merged**